src/System.Diagnostics.PerformanceCounter/src/System/Diagnostics/PerformanceDataRegistryKey.cs

```diff
@@ -0,0 +1,119 @@
+
+            if (ret == Interop.Errors.ERROR_DLL_INIT_FAILED)
+            {
+                // return value indicates an error occurred
+                throw new ArgumentException(SR.Format(SR.Arg_DllInitFailure, machineName));
+            }
+
+            if (ret != 0)
+            {
+                Win32Error(ret, null);
+            }
+
+            if (foreignHKey.IsInvalid)
+            {
+                // return value indicates an error occurred
+                throw new ArgumentException(SR.Format(SR.Arg_RegKeyNoRemoteConnect, machineName));
+            }
+
+            return new PerformanceDataRegistryKey(foreignHKey);
+        }
+
+        public static PerformanceDataRegistryKey OpenLocal()
+        {
+            var key = new SafeRegistryHandle(new IntPtr(PerformanceData), true);
+            return new PerformanceDataRegistryKey(key);
+        }
+
+        public byte[] GetValue(string name, bool usePool)
+        {
+            int size = 65000;
+            int sizeInput = size;
+
+            int ret;
+            int type = 0;
+            byte[] blob = CreateBlob(size, usePool);
+            while (Interop.Errors.ERROR_MORE_DATA == (ret = Interop.Advapi32.RegQueryValueEx(_hkey, name, null, ref type, blob, ref sizeInput)))
+            {
+                if (size == int.MaxValue)
+                {
+                    // ERROR_MORE_DATA was returned however we cannot increase the buffer size beyond Int32.MaxValue
+                    Win32Error(ret, name);
+                }
+                else if (size > (int.MaxValue / 2))
+                {
+                    // at this point in the loop "size * 2" would cause an overflow
+                    size = int.MaxValue;
+                }
+                else
+                {
+                    size *= 2;
+                }
+                sizeInput = size;
+
+                ReleaseBlob(blob, usePool);
+                blob = CreateBlob(size, usePool);
+            }
+
+            if (ret != 0)
+            {
+                Win32Error(ret, name);
+            }
+
+            return blob;
+        }
+
+        public void ReleaseBlob(byte[] blob, bool usePool = true)
+        {
+            if (usePool)
+            {
+                ArrayPool<byte>.Shared.Return(blob);
+            }
+        }
+
+        private byte[] CreateBlob(int size, in bool usePool)
+        {
+            return usePool
+                ? ArrayPool<byte>.Shared.Rent(size)
+                : new byte[size];
+        }
+
+        private static void Win32Error(in int errorCode, string name)
+        {
+            if (errorCode == Interop.Errors.ERROR_ACCESS_DENIED)
+            {
+                throw new UnauthorizedAccessException(SR.Format(SR.UnauthorizedAccess_RegistryKeyGeneric_Key, name));
+            }
+
+            throw new IOException(Interop.Kernel32.GetMessage(errorCode), errorCode);
+        }
+
+        public void Close()
```

Comment:
Should this class be disposable, like a regular registry key, and this be Dispose()?

Reply:
I think it should.

Reply:
It should also be explicitly disposing `_hkey` and not leaving it to the finalizer.

Reply:
Instead of explicitly calling `RegCloseKey`. Wouldn't the handle end up getting closed twice?

Reply:
Ah, you pass false to `ownsHandle` when creating the SafeHandle. If you pass true, can you dispose it explicitly here?

Reply:
I'll implement `IDisposable` here, but, as I understand, current behavior (even before my changes) don't call `Close`/`Dispose` on `RegistryKey` even on dispose remote `PerformanceCounter`. This method called only when Create/Remove category or `PerofrmanceCounter.CloseSharedResources()`. I think we should create another issue related with it?

Reply:
@jkotas if a type owns a disposable resource, and has a `Close()` method to dispose it (as here) is it important that we make that type actually disposable (and presumably rename `Close()`) or does it not matter? From the GC point of view, it does not matter -- if `Close()` is called, the resource is disposed, if it isn't, the finalizer cleans it up, either way. Perhaps it is just a style issue.